EIP-009-js-pdk-interface-definition.md

```diff
@@ -0,0 +1,135 @@
+
+The programmer will define their interface in a typescript file `interface.d.ts`:
+
+```typescript
+// all extism exports have type () -> i32
+interface Exports {
+  myExport(): I32;
+}
+
+// this interface is optional and types are  defined by you
+// to match your host functions
+interface Imports {
+  myHostFunc1(p: I64, q: I32): I64;
+  myHostFunc2(p: I64): I64;
+}
+```
+
+Step 1 of the CLI will be to generate 2 shim modules `export-shim.wasm` and `import-shim.wasm` from this interface.
+The export shim needs to generate some thunk functions as well as import `__invoke` from the core module (which gets
+linked later in the pipeline). I spiked a working [prototype of this here](https://gist.github.com/bhelx/41fba8959fe7738a23cd750983341216).
+
+`export-shim.wasm` will be linked with `core.wasm` using [wasm-merge](https://github.com/WebAssembly/binaryen#wasm-merge)
+which is kind of like a linker but a little bit higher level. I may also consider using `wasm-ld`
+but this seems to work for me and is flexible.
+
+We then wasm-merge the import shim which gives us a unified interface for indirectly calling import functions.
+
+Out of thhis merge comes the final module, but it's yet to be wizened. This is similar to the process
+wer have now. We will run the module and eval the user's plugin js code in the egine. Then we freeze
+and dump out to the final wasm after a few passes from wasm-opt.
+
+### Rust to JS Bridge
+
+Suppose we have 2 host functions and one export. The shim module that is generated
+should look something like this (TODO this is still not accurate yet):
+
+export-shim.wasm:
+
+```wat
+(module
+  ; included by us for invoking the js runtime
+  (import "coremod" "__invoke" (func (;0;) (type 0)))
+
+  ; the generated thunk for the export
+  (func (;1;) (type 0) (param i32) (result i32)
+    local.get 0
+    call 0)
+  (export "myExport" (func 1)))
+```
+
+For exports, we can leave the behavior the same. For each export we will generate
+a thunk function that calls `__invoke`.
+
+For imports we have a bit of a challenge. The JS code, which is executed in the core
+module, needs to be able to invoke these host functions.
+
+The JS code (and underlying rust and c code) doesn't know about these functions, their names or their function
+indexes at compile time. That's why the import shim will expose a call-indirect to the host functions.
+
+TODO: this means all types need to be the same, perhaps we need to also expose the type parameter?
+
+import-shim.wasm:
+
+```wat
+(module
+  (type $int2int (func (param i32) (result i32)))
+
+  (import "coremode" "myHostFunc1" (func $myHostFunc1 (type $int2int)))
+  (import "coremode" "myHostFunc2" (func $myHostFunc2 (type $int2int)))
+
+  (table 2 funcref)
+  (elem (i32.const 0) $myHostFunc1 $myHostFunc2)
+
+  (func $callHostFunc (type $int2int)
+    local.get 0
+    (call_indirect (type $int2int) (i32.const 0))
+  )
+  (export "__invokeHostFunc" (func $callHostFunc))
```

Comment:
(We talked offline a bit about this, but noting it here for posterity) dispatching based on the different host import type signatures might be a little tricky (but not insurmountable!)